google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java

```diff
@@ -0,0 +1,292 @@
+    while (repos.hasMoreElements()) {
+      TreeNode repo = (TreeNode) repos.nextElement();
+
+      if (repo instanceof RepositoryModelItem
+          && getText().equals(((RepositoryModelItem) repo).getRepositoryId())) {
+        return getText();
+      }
+    }
+
+    return null;
+  }
+
+  public void setCloudProject(String cloudProject) {
+    this.cloudProject = cloudProject;
+  }
+
+  public void setUser(CredentialedUser user) {
+    this.user = user;
+  }
+
+  @Override
+  public void showPopup(RelativePoint showTarget) {
+    if (user != null) {
+      if (popup == null || popup.isDisposed()) {
+        panel = new RepositoryPanel(cloudProject);
+
+        ComponentPopupBuilder popupBuilder = JBPopupFactory.getInstance()
+            .createComponentPopupBuilder(panel, null);
+        popup = popupBuilder.createPopup();
+      }
+      if (!popup.isVisible()) {
+        popup.show(showTarget);
+      }
+    } else {
+      panel = new ProjectNotSelectedPanel();
+
+      ComponentPopupBuilder popupBuilder = JBPopupFactory.getInstance()
+            .createComponentPopupBuilder(panel, null);
+      popup = popupBuilder.createPopup();
+      popup.show(showTarget);
+    }
+  }
+
+  @Override
+  public void hidePopup() {
+    if (isPopupVisible()) {
+      popup.closeOk(null);
+    }
+  }
+
+  @Override
+  public boolean isPopupVisible() {
+    return popup != null && !popup.isDisposed() && popup.isVisible();
+  }
+
+  @Override
+  protected CustomizableComboBoxPopup getPopup() {
+    return this;
+  }
+
+  /**
+   * Returns the width of the repository select popup to show. The popup show have a minimum width
+   * and then expand with the repository textfield if the user manually stretches the dialog.
+   */
+  private int getPopupWidth() {
+    int actualWidth = RepositorySelector.this.getTextField().getWidth();
+    return Math.max(SELECTOR_WIDTH, actualWidth);
+  }
+
+  @Override
+  protected int getPreferredPopupHeight() {
+    return SELECTOR_HEIGHT;
+  }
+
+  private class ProjectNotSelectedPanel extends JPanel {
+
+    private static final int HEIGHT = 30;
+
+    ProjectNotSelectedPanel() {
+      setPreferredSize(new Dimension(RepositorySelector.this.getTextField().getWidth(), HEIGHT));
+      JLabel warning = new JBLabel();
+      warning.setFont(new Font(getFont().getFontName(), Font.ITALIC, getFont().getSize()));
+      warning.setText(GctBundle.message("cloud.repository.selector.missing.project.error"));
+      add(warning);
+    }
+  }
+
+  private class RepositoryPanel extends JPanel {
+
+    private JTree repositoryTree;
+    private DefaultMutableTreeNode projectRootNode;
+    private DefaultTreeModel treeModel;
+    private ProjectRepositoriesModelItem repositories;
+
+    RepositoryPanel(String cloudProject) {
+      repositories
+          = new ProjectRepositoriesModelItem(cloudProject, user);
+      projectRootNode = new DefaultMutableTreeNode("root");
```

Comment:
no, the string "root" has no significance, its there just for clarity. I think the tree will behave differently if I were to pass in null.